client/apps/edit/components/section_image_set/index.coffee

```diff
@@ -0,0 +1,263 @@
+#
+# Image Set section that allows uploading a mix of images and artworks
+#
+
+# Using `try` here b/c Scribe is an AMD module that doesn't play nice when
+# requiring it for testing in node.
+try
+  Scribe = require 'scribe-editor'
+  scribePluginToolbar = require 'scribe-plugin-toolbar'
+  scribePluginSanitizer = require '../../lib/sanitizer.coffee'
+  scribePluginLinkTooltip = require 'scribe-plugin-enhanced-link-tooltip'
+_ = require 'underscore'
+gemup = require 'gemup'
+React = require 'react'
+toggleScribePlaceholder = require '../../lib/toggle_scribe_placeholder.coffee'
+sd = require('sharify').data
+icons = -> require('./icons.jade') arguments...
+Autocomplete = require '../../../../components/autocomplete/index.coffee'
+Artwork = require '../../../../models/artwork.coffee'
+{ div, section, h1, h2, span, img, header, input, nav, a, button, p, ul, li, strong } = React.DOM
+{ crop, resize, fill } = embedly = require('embedly-view-helpers')(sd.EMBEDLY_KEY)
+
+module.exports = React.createClass
+
+  getInitialState: ->
+    images: @props.section.get('images') or []
+    progress: null
+
+  componentDidMount: ->
+    @attachScribe()
+    @setupAutocomplete()
+    @showPreviewImages()
+
+  componentDidUpdate: ->
+    @attachScribe()
+
+  componentWillUnmount: ->
+    @autocomplete.remove()
+
+  onStateChange: ->
+    @props.section.set images: @state.images if @state.images.length > 0
+    @showPreviewImages()
+    @forceUpdate()
+
+  setupAutocomplete: ->
+    $el = $(@refs.autocomplete.getDOMNode())
+    @autocomplete = new Autocomplete
+      url: "#{sd.ARTSY_URL}/api/search?q=%QUERY"
+      el: $el
+      filter: (res) ->
+        vals = []
+        for r in res._embedded.results
+          if r.type == 'Artwork'
+            id = r._links.self.href.substr(r._links.self.href.lastIndexOf('/') + 1)
+            vals.push
+              id: id
+              value: r.title
+              thumbnail: r._links.thumbnail.href
+        return vals
+      templates:
+        suggestion: (data) ->
+          """
+            <div class='esis-suggestion' \
+                 style='background-image: url(#{data.thumbnail})'>
+            </div>
+            #{data.value}
+          """
+      selected: @onSelect
+    _.defer -> $el.focus()
+
+  onSelect: (e, selected) ->
+    new Artwork(id: selected.id).fetch
+      success: (artwork) =>
+        @state.images.push artwork.denormalized()
```

Comment:
You shouldn't be mutating state directly here. Make sure to always use `setState` instead. e.g.

````coffeescript
newImages = @state.images.concat [artwork.denormalized()]
@setState images: newImages
````